components/script/dom/servoparser/async_html.rs

```diff
@@ -20,218 +21,249 @@ use dom::htmltemplateelement::HTMLTemplateElement;
-    unsafe fn trace(&self, trc: *mut JSTracer) {
-        struct Tracer(*mut JSTracer);
-        let tracer = Tracer(trc);
-
-        impl HtmlTracer for Tracer {
-            type Handle = ParseNode;
-            #[allow(unrooted_must_root)]
-            fn trace_handle(&self, node: &ParseNode) {
-                unsafe { node.trace(self.0); }
-            }
-        }
-
-        let tree_builder = &self.sink;
-        tree_builder.trace_handles(&tracer);
-        tree_builder.sink.trace(trc);
-    }
-}
-
 type ParseNodeId = usize;
 
-#[derive(JSTraceable, Clone, HeapSizeOf)]
+#[derive(Clone, HeapSizeOf, JSTraceable)]
 pub struct ParseNode {
     id: ParseNodeId,
     qual_name: Option<QualName>,
 }
 
-#[derive(JSTraceable, HeapSizeOf)]
-struct ParseNodeData {
-    contents: Option<ParseNode>,
-    is_integration_point: bool,
+#[derive(HeapSizeOf, JSTraceable)]
+enum NodeOrText {
+    Node(ParseNode),
+    Text(String),
 }
 
-impl Default for ParseNodeData {
-    fn default() -> ParseNodeData {
-        ParseNodeData {
-            contents: None,
-            is_integration_point: false,
-        }
-    }
+#[derive(HeapSizeOf, JSTraceable)]
+struct Attribute {
+    name: QualName,
+    value: String,
 }
 
+#[derive(HeapSizeOf, JSTraceable)]
 enum ParseOperation {
     GetTemplateContents(ParseNodeId, ParseNodeId),
-    CreateElement(ParseNodeId, QualName, Vec<Attribute>),
-    CreateComment(StrTendril, ParseNodeId),
+    CreateElement(ParseNodeId, QualName, Vec<Attribute>, u64),
+    CreateComment(String, ParseNodeId),
     // sibling, node to be inserted
-    AppendBeforeSibling(ParseNodeId, NodeOrText<ParseNode>),
+    AppendBeforeSibling(ParseNodeId, NodeOrText),
     // parent, node to be inserted
-    Append(ParseNodeId, NodeOrText<ParseNode>),
-    AppendDoctypeToDocument(StrTendril, StrTendril, StrTendril),
+    Append(ParseNodeId, NodeOrText),
+    AppendDoctypeToDocument(String, String, String),
     AddAttrsIfMissing(ParseNodeId, Vec<Attribute>),
     RemoveFromParent(ParseNodeId),
     MarkScriptAlreadyStarted(ParseNodeId),
     ReparentChildren(ParseNodeId, ParseNodeId),
     AssociateWithForm(ParseNodeId, ParseNodeId),
-    CreatePI(ParseNodeId, StrTendril, StrTendril),
+    CreatePI(ParseNodeId, String, String),
     Pop(ParseNodeId),
+    SetQuirksMode(
+        #[ignore_heap_size_of = "Defined in style"]
+        ServoQuirksMode
+    ),
 }
 
-#[derive(JSTraceable, HeapSizeOf)]
+#[derive(HeapSizeOf)]
+enum FromHtmlTokenizerMsg {
+    Result(#[ignore_heap_size_of = "Defined in html5ever"]TokenizerResult<ParseNode>)
+}
+
+#[derive(HeapSizeOf)]
+enum ToHtmlTokenizerMsg {
+    Feed(#[ignore_heap_size_of = "Defined in html5ever"]Arc<Mutex<Vec<SendTendril<UTF8>>>>),
+    End,
+    SetPlainTextState,
+}
+
+#[derive(HeapSizeOf)]
+enum FromSinkMsg {
+    ProcessOperation(ParseOperation),
+    IsSameTree(ParseNodeId, ParseNodeId),
+    HasParentNode(ParseNodeId),
+}
+
+#[derive(HeapSizeOf)]
+enum ToSinkMsg {
```

Comment:
The messages sent from the main thread Tokenizer to the Sink are the results of the synchronous operations we perform. Only two sync operations exist.